File: content/examples/other/leetcode/problems/clone-graph/example-1/solution.py
class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


def clone_graph_dfs(node):
    if node is None:
        return None
    seen = {}

    def dfs(cur):
        if cur in seen:
            return seen[cur]
        copy = Node(cur.val, [])
        seen[cur] = copy
        for nxt in cur.neighbors:
            copy.neighbors.append(dfs(nxt))
        return copy

    return dfs(node)


if __name__ == "__main__":
    a = Node(1)
    b = Node(2)
    a.neighbors = [b]
    b.neighbors = [a]
    cloned = clone_graph_dfs(a)
    print(cloned.val, cloned.neighbors[0].val)

File: content/examples/other/leetcode/problems/clone-graph/example-1/Solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Solution {
  static class Node {
    public int val;
    public List<Node> neighbors;

    public Node(int val) {
      this.val = val;
      this.neighbors = new ArrayList<>();
    }
  }

  static Node cloneGraphDFS(Node node) {
    if (node == null) return null;
    Map<Node, Node> seen = new HashMap<>();
    return dfs(node, seen);
  }

  static Node dfs(Node cur, Map<Node, Node> seen) {
    if (seen.containsKey(cur)) return seen.get(cur);
    Node copy = new Node(cur.val);
    seen.put(cur, copy);
    for (Node nxt : cur.neighbors) copy.neighbors.add(dfs(nxt, seen));
    return copy;
  }

  public static void main(String[] args) {
    Node a = new Node(1);
    Node b = new Node(2);
    a.neighbors.add(b);
    b.neighbors.add(a);
    Node cloned = cloneGraphDFS(a);
    System.out.println(cloned.val + " " + cloned.neighbors.get(0).val);
  }
}
